File: cc/mac/aes_cmac_key_manager.cc
// Copyright 2017 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
///////////////////////////////////////////////////////////////////////////////

#include "tink/mac/aes_cmac_key_manager.h"

#include <map>

#include "absl/strings/string_view.h"
#include "tink/key_manager.h"
#include "tink/mac.h"
#include "tink/subtle/aes_cmac_boringssl.h"
#include "tink/subtle/random.h"
#include "tink/util/enums.h"
#include "tink/util/errors.h"
#include "tink/util/protobuf_helper.h"
#include "tink/util/status.h"
#include "tink/util/statusor.h"
#include "tink/util/validation.h"
#include "proto/aes_cmac.pb.h"
#include "proto/tink.pb.h"

namespace crypto {
namespace tink {

using crypto::tink::util::Enums;
using crypto::tink::util::Status;
using crypto::tink::util::StatusOr;
using google::crypto::tink::AesCmacKey;
using google::crypto::tink::AesCmacKeyFormat;
using google::crypto::tink::AesCmacParams;
using google::crypto::tink::KeyData;

constexpr uint32_t AesCmacKeyManager::kVersion;
// Due to https://www.math.uwaterloo.ca/~ajmeneze/publications/tightness.pdf, we
// only allow key sizes of 256 bit.
constexpr int kKeySizeInBytes = 32;
constexpr int kMaxTagSizeInBytes = 16;
constexpr int kMinTagSizeInBytes = 10;

class AesCmacKeyFactory : public KeyFactoryBase<AesCmacKey, AesCmacKeyFormat> {
 public:
  AesCmacKeyFactory() {}

  KeyData::KeyMaterialType key_material_type() const override {
    return KeyData::SYMMETRIC;
  }

 protected:
  StatusOr<std::unique_ptr<AesCmacKey>> NewKeyFromFormat(
      const AesCmacKeyFormat& cmac_key_format) const override;
};

StatusOr<std::unique_ptr<AesCmacKey>> AesCmacKeyFactory::NewKeyFromFormat(
    const AesCmacKeyFormat& cmac_key_format) const {
  Status status = AesCmacKeyManager::Validate(cmac_key_format);
  if (!status.ok()) return status;
  auto cmac_key = absl::make_unique<AesCmacKey>();
  cmac_key->set_version(AesCmacKeyManager::kVersion);
  cmac_key->set_key_value(
      subtle::Random::GetRandomBytes(cmac_key_format.key_size()));
  *cmac_key->mutable_params() = cmac_key_format.params();
  return absl::implicit_cast<StatusOr<std::unique_ptr<AesCmacKey>>>(
      std::move(cmac_key));
}

AesCmacKeyManager::AesCmacKeyManager()
    : key_factory_(new AesCmacKeyFactory()) {}

uint32_t AesCmacKeyManager::get_version() const { return kVersion; }

const KeyFactory& AesCmacKeyManager::get_key_factory() const {
  return *key_factory_;
}

StatusOr<std::unique_ptr<Mac>> AesCmacKeyManager::GetPrimitiveFromKey(
    const AesCmacKey& cmac_key) const {
  Status status = Validate(cmac_key);
  if (!status.ok()) return status;
  auto cmac_result = subtle::AesCmacBoringSsl::New(
      cmac_key.key_value(), cmac_key.params().tag_size());
  if (!cmac_result.ok()) return cmac_result.status();
  return std::move(cmac_result.ValueOrDie());
}

// static
Status AesCmacKeyManager::Validate(const AesCmacParams& params) {
  if (params.tag_size() < kMinTagSizeInBytes) {
    return ToStatusF(util::error::INVALID_ARGUMENT,
                     "Invalid AesCmacParams: tag_size %d is too small.",
                     params.tag_size());
  }
  if (params.tag_size() > kMaxTagSizeInBytes) {
    return ToStatusF(util::error::INVALID_ARGUMENT,
                     "Invalid AesCmacParams: tag_size %d is too big.",
                     params.tag_size());
  }
  return Status::OK;
}

// static
Status AesCmacKeyManager::Validate(const AesCmacKey& key) {
  Status status = ValidateVersion(key.version(), kVersion);
  if (!status.ok()) return status;
  if (key.key_value().size() != kKeySizeInBytes) {
    return ToStatusF(util::error::INVALID_ARGUMENT,
                     "Invalid AesCmacKey: key_value wrong length.");
  }
  return Validate(key.params());
}

// static
Status AesCmacKeyManager::Validate(const AesCmacKeyFormat& key_format) {
  if (key_format.key_size() != kKeySizeInBytes) {
    return ToStatusF(util::error::INVALID_ARGUMENT,
                     "Invalid AesCmacKeyFormat: invalid key_size.");
  }
  return Validate(key_format.params());
}

}  // namespace tink
}  // namespace crypto

File: go/mac/aes_cmac_key_manager.go
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

package mac

import (
	"errors"
	"fmt"

	"google.golang.org/protobuf/proto"
	"github.com/google/tink/go/keyset"
	"github.com/google/tink/go/mac/subtle"
	"github.com/google/tink/go/subtle/random"
	cmacpb "github.com/google/tink/go/proto/aes_cmac_go_proto"
	tinkpb "github.com/google/tink/go/proto/tink_go_proto"
)

const (
	cmacKeyVersion = 0
	cmacTypeURL    = "type.googleapis.com/google.crypto.tink.AesCmacKey"
)

var errInvalidCMACKey = errors.New("aes_cmac_key_manager: invalid key")
var errInvalidCMACKeyFormat = errors.New("aes_cmac_key_manager: invalid key format")

// cmacKeyManager generates new AES-CMAC keys and produces new instances of AES-CMAC.
type aescmacKeyManager struct{}

// Primitive constructs a AES-CMAC instance for the given serialized CMACKey.
func (km *aescmacKeyManager) Primitive(serializedKey []byte) (interface{}, error) {
	if len(serializedKey) == 0 {
		return nil, errInvalidCMACKey
	}
	key := new(cmacpb.AesCmacKey)
	if err := proto.Unmarshal(serializedKey, key); err != nil {
		return nil, errInvalidCMACKey
	}
	if err := km.validateKey(key); err != nil {
		return nil, err
	}
	cmac, err := subtle.NewAESCMAC(key.KeyValue, key.Params.TagSize)
	if err != nil {
		return nil, err
	}
	return cmac, nil
}

// NewKey generates a new AesCmacKey according to specification in the given AesCmacKeyFormat.
func (km *aescmacKeyManager) NewKey(serializedKeyFormat []byte) (proto.Message, error) {
	if len(serializedKeyFormat) == 0 {
		return nil, errInvalidCMACKeyFormat
	}
	keyFormat := new(cmacpb.AesCmacKeyFormat)
	if err := proto.Unmarshal(serializedKeyFormat, keyFormat); err != nil {
		return nil, errInvalidCMACKeyFormat
	}
	if err := km.validateKeyFormat(keyFormat); err != nil {
		return nil, fmt.Errorf("aes_cmac_key_manager: invalid key format: %s", err)
	}
	keyValue := random.GetRandomBytes(keyFormat.KeySize)
	return &cmacpb.AesCmacKey{
		Version:  cmacKeyVersion,
		Params:   keyFormat.Params,
		KeyValue: keyValue,
	}, nil
}

// NewKeyData generates a new KeyData according to specification in the given
// serialized AesCmacKeyFormat. This should be used solely by the key management API.
func (km *aescmacKeyManager) NewKeyData(serializedKeyFormat []byte) (*tinkpb.KeyData, error) {
	key, err := km.NewKey(serializedKeyFormat)
	if err != nil {
		return nil, err
	}
	serializedKey, err := proto.Marshal(key)
	if err != nil {
		return nil, errInvalidCMACKeyFormat
	}

	return &tinkpb.KeyData{
		TypeUrl:         cmacTypeURL,
		Value:           serializedKey,
		KeyMaterialType: tinkpb.KeyData_SYMMETRIC,
	}, nil
}

// DoesSupport checks whether this KeyManager supports the given key type.
func (km *aescmacKeyManager) DoesSupport(typeURL string) bool {
	return typeURL == cmacTypeURL
}

// TypeURL returns the type URL of keys managed by this KeyManager.
func (km *aescmacKeyManager) TypeURL() string {
	return cmacTypeURL
}

// validateKey validates the given AesCmacKey. It only validates the version of the
// key because other parameters will be validated in primitive construction.
func (km *aescmacKeyManager) validateKey(key *cmacpb.AesCmacKey) error {
	err := keyset.ValidateKeyVersion(key.Version, cmacKeyVersion)
	if err != nil {
		return fmt.Errorf("aes_cmac_key_manager: invalid version: %s", err)
	}
	keySize := uint32(len(key.KeyValue))
	return subtle.ValidateCMACParams(keySize, key.Params.TagSize)
}

// validateKeyFormat validates the given AesCmacKeyFormat
func (km *aescmacKeyManager) validateKeyFormat(format *cmacpb.AesCmacKeyFormat) error {
	if format.Params == nil {
		return fmt.Errorf("null AES-CMAC params")
	}
	return subtle.ValidateCMACParams(format.KeySize, format.Params.TagSize)
}
